src/main/java/com/yahoo/bullet/pubsub/PubSubMessageSerDe.java

```diff
@@ -0,0 +1,81 @@
+/*
+ *  Copyright 2021 Yahoo Inc.
+ *  Licensed under the terms of the Apache License, Version 2.0.
+ *  See the LICENSE file associated with the project for terms.
+ */
+package com.yahoo.bullet.pubsub;
+
+import com.yahoo.bullet.common.BulletConfig;
+import com.yahoo.bullet.query.Query;
+
+import java.io.Serializable;
+
+/**
+ * This allows you to hook into the {@link PubSubMessage} serialization and deserialization. In particular, if you want
+ * to customize what is done to the message before it is sent to or read in the backend. You can implement a custom
+ * version of this to change what the {@link PubSubMessage} contains (see {@link #toMessage(PubSubMessage)} and invert
+ * it when it is read back (see {@link #fromMessage(PubSubMessage)}.
+ *
+ * Please note that this is not intended to be applied to messages read from the backend or for signals sent to the
+ * backend. You can for example, use this if you want to customize what is stored in storage or to lazily create a
+ * {@link Query} in the backend.
+ */
+public abstract class PubSubMessageSerDe implements Serializable {
+    private static final long serialVersionUID = 5352288558800960763L;
+
+    protected BulletConfig config;
+
+    /**
+     * Constructor.
+     *
+     * @param config The {@link BulletConfig} to configure this class.
+     */
+    public PubSubMessageSerDe(BulletConfig config) {
+        this.config = config;
+    }
+
+    /**
+     * A helper to submit a query by converting it to a standard {@link PubSubMessage} and running it through
+     * {@link #toMessage(PubSubMessage)}.
+     *
+     * @param id The ID for the query.
+     * @param query The {@link Query} object to create a {@link PubSubMessage}
+     * @param queryString The BQL for the query.
+     * @return A converted {@link PubSubMessage}.
+     */
+    public PubSubMessage toMessage(String id, Query query, String queryString) {
+        return toMessage(new PubSubMessage(id, query, new Metadata(null, queryString)));
+    }
+
+    /**
+     * Takes a {@link PubSubMessage} and returns the new format of the message. Note that it is allowed to modify the
```

Comment:
how likely is it for us to actually modify the message rather than creating a new pubsubmessage with different content?

Reply:
I can see how we might need to modify the message. I'll just add a setter.